### PR TITLE
i.biomass: added test file for i.biomass module 

### PR DESCRIPTION
**Pull request by jayneel-shah18**
This PR introduces a suite of tests for the `i.biomass` GRASS GIS module, covering a range of scenarios, including functionality validation, property testing, edge cases, and performance evaluation.

**Key updates in this PR include:**
- **Functionality:** Verify output raster existence and dimensions.
- **Properties:** Test linearity and latitude dependency.
- **Edge Cases:** Validate behavior with zero and extreme input values.
- **Performance:** Test large-scale processing.

**Test Case Additions:**
- **Basic Functionality:**
  - Validate the creation and existence of the biomass output raster.

- **Advanced Properties:**
  - Linearity: Verify output scales proportionally with input.
  - Latitude Dependency: Ensure biomass values vary reasonably with latitude.

- **Edge Cases:**
  - Handle all-zero input rasters.
  - Test extreme values for all parameters.

- **Performance:**
  - Assess processing of large input rasters.

### CODE REVIEW

File: imagery/i.biomass/testsuite/test_i_biomass.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+        )
+        self.assertRasterExists(self.output_raster)
+
+        output_values_max = array.array(self.output_raster)
+        self.assertTrue(
+            np.all(np.isfinite(output_values_max)),
+            "Biomass output contains non-finite values.",
+        )
+
+    def test_biomass_large_processing(self):
```
Comment (petrasovaa) [inline imagery/i.biomass/testsuite/test_i_biomass.py:202]: I think this test is not needed, it's simple tool and it could slow it down. BTW, the tests in CI run in EPSG:3358, so the region bounds here are not in degrees but meters.

File: imagery/i.biomass/testsuite/test_i_biomass.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+        cls.runModule(
+            "g.remove",
+            type="raster",
+            name=",".join(rasters_to_remove),
+            flags="f",
+            quiet=True,
+        )
+        cls.del_temp_region()
+
+    def test_biomass_with_zero_inputs(self):
```
Comment (petrasovaa) [inline imagery/i.biomass/testsuite/test_i_biomass.py:70]: I don't think this is particularly useful test. Instead, do a regression test to make sure the values stay the same.

File: imagery/i.biomass/testsuite/test_i_biomass.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    }
+    output_raster = "biomass_output"
+
+    @classmethod
+    def setUpClass(cls):
+        """Set up input rasters and configure test environment."""
+        cls.use_temp_region()
+        cls.runModule("g.region", n=10, s=0, e=10, w=0, rows=10, cols=10)
+
+        cls.runModule(
```
Comment (petrasovaa) [inline imagery/i.biomass/testsuite/test_i_biomass.py:26]: Maybe make a function for creating all rasters?

Also note the order of the tests is not guaranteed.

**Review by petrasovaa — CHANGES_REQUESTED** [review]

File: imagery/i.biomass/testsuite/test_i_biomass.py (hunk 1 above)
Comment (jayneel-shah18) [reply imagery/i.biomass/testsuite/test_i_biomass.py]: Thanks for the feedback! That makes sense—I have removed the test from the suite. I appreciate the note about the coordinate system and will keep that in mind for future tests.

File: imagery/i.biomass/testsuite/test_i_biomass.py (hunk 2 above)
Comment (jayneel-shah18) [reply imagery/i.biomass/testsuite/test_i_biomass.py]: Thank you for the suggestion! I have replaced the test with a regression test to ensure the values remain consistent.

File: imagery/i.biomass/testsuite/test_i_biomass.py (hunk 3 above)
Comment (jayneel-shah18) [reply imagery/i.biomass/testsuite/test_i_biomass.py]: I have refactored the raster creation into a separate function `_create_input_rasters()` to improve readability and maintainability.

File: imagery/i.biomass/testsuite/test_i_biomass.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+        self.assertRasterExists(self.output_raster)
+        self.assertRasterExists(reference_raster)
+
+        output_values = array.array(self.output_raster)
+        reference_values = array.array(reference_raster)
+
+        self.assertTrue(
+            np.allclose(output_values, reference_values, atol=2.0),
+            "Biomass raster values should match the reference values",
+        )
```
Comment (petrasovaa) [inline imagery/i.biomass/testsuite/test_i_biomass.py:97]: That's not what I had in mind. You are computing the reference raster with each test run, but the idea is to compute the reference at the moment you are writing the test and then this reference is considered to be correct, the values are fixed in the test. So when somebody later changes the tool, resulting in different results, the test will fail and you will know that there is potentially a newly introduced bug. See for example r.viewshed test. It is often enough to compute univariate statistics with r.univar and then test those statistics, not cell-by-cell comparison.

File: imagery/i.biomass/testsuite/test_i_biomass.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+        expected_stats = (
+            "min=4.51592674628609\n"
+            "max=75.2654457714349\n"
+            "mean=33.1167961394313\n"
+            "stddev=18.5128518411928"
+        )
+
+        self.assertRasterFitsUnivar(
+            raster=self.output_raster, reference=expected_stats, precision=1e-8
```
Comment (petrasovaa) [inline imagery/i.biomass/testsuite/test_i_biomass.py:95]: In other tests we use slightly lower precision which is sufficient. Also you can reduce the number of decimal places in the stats.
```suggestion
            raster=self.output_raster, reference=expected_stats, precision=1e-6
```

File: imagery/i.biomass/testsuite/test_i_biomass.py [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    def setUpClass(cls):
+        """Set up input rasters and configure test environment."""
+        cls.use_temp_region()
+        cls.runModule("g.region", n=10, s=0, e=10, w=0, rows=10, cols=10)
+        cls._create_input_rasters()
+
+    @classmethod
+    def _create_input_rasters(cls):
+        """Helper method to create all input rasters with defined expressions."""
+        expressions = {
```
Comment (petrasovaa) [inline imagery/i.biomass/testsuite/test_i_biomass.py:30]: This could be perhaps the method's parameter and then you can use this method in `test_biomass_ecological_range` and maybe other places.

**Review by petrasovaa — CHANGES_REQUESTED** [review]
The test functions can be executed in undefined order, so in the setUpClass you create the initial inputs, but the individual test functions modify them, so make sure that's not  going to be a problem.

File: imagery/i.biomass/testsuite/test_i_biomass.py (hunk 5 above)
Comment (jayneel-shah18) [reply imagery/i.biomass/testsuite/test_i_biomass.py]: I have changed to six decimal places and also reduced the number of decimal places in the stats.

File: imagery/i.biomass/testsuite/test_i_biomass.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        input_expressions = {
+            "fpar": "col() * 0.1",
+            "lightuse_eff": "row() * 0.1",
+            "latitude": "45.0",
+            "dayofyear": "150",
+            "transmissivity": "0.75",
+            "water": "0.8",
+        }
+        self._create_input_rasters(input_expressions)
+
```
Comment (petrasovaa) [inline imagery/i.biomass/testsuite/test_i_biomass.py:37]: I think this `setUp` is not a bad idea but given many tests are anyway creating their own layers, this is being run wastefully, so I would compute the data at the beginning of each test. We are trying to make the tests run fast.

**Review by petrasovaa — CHANGES_REQUESTED** [review]

File: imagery/i.biomass/testsuite/test_i_biomass.py (hunk 7 above)
Comment (echoix) [reply imagery/i.biomass/testsuite/test_i_biomass.py:37]: It's 5 seconds for now, some others in imagery take 1 second, some others take more.

Comment (jayneel-shah18) [reply imagery/i.biomass/testsuite/test_i_biomass.py]: Thank you for the feedback! I have restructured the tests to address the feedback about efficiency. The `setUp` method has been removed, and input rasters are now created directly within each test as needed avoiding redundant computations and ensuring faster test execution.

**Review by petrasovaa — APPROVED** [review]
